[2.0.x] Migrate Marlin to a new TMC library

teemuatlut opened the pull request:
I'm starting the process to migrate to a new TMC library that aims to support more driver models.
It might be a bit early but I'd like to get a review going to make things go smoothly.

Here's a general plan for the rest of the year. Ticked boxes mark the changes introduced in this PR.
- [x] Move to new universal TMCStepper library
- [x] Implement inherited Marlin TMC classes
- [ ] Support TMC2660
- [ ] Support TMC5160
- [ ] Add sanity checking for driver errors
- [ ] Greatly reduce communication attempts for monitoring
- [ ] Improve TMC error verbosity with better error messages
- [ ] Add communication checking on boot
- [ ] Add the ability to change chopper timings (spreadCycle)
- [ ] Tweak default chopper settings for quieter spreadCycle operation
- [ ] Support Trinamic coolStep technology
- [ ] Improve `M122` with fetching raw register data
- [ ] Add TMC section to the LCD
- [ ] Support TMC5130
- [ ] Support TMC2224
- [ ] Add `SOFTWARE_ENABLE` for all smart drivers (Request from Panucatt)

---

Regarding this PR.

A more universal library allows me to support more chips without creating a completely new library for each and every one of them. The maintenance would quickly become too great when the number of supported models is more than a couple.
The current aim is to support TMC2130, TMC2208, TMC2224, TMC5130, TMC5160 and TMC2660.
The library itself should be in pretty good condition but as always, it is mostly untested at this point.

Implementing a child class within Marlin allows us to attach certain functions to the stepper objects so they don't have to be passed around as parameters, or be tracked as indexes to an array. This PR primarily uses this for printing out the axis labels and otpw counter values.

Basically this PR forms the base for many of the new features, functions and fixes.
I ask that the PRs don't get squashed together as I've tried to organize the commits thematically and it'll make it easier to debug in case of any issues.

teemuatlut committed:
Implement inherited TMCMarlin class

teemuatlut committed:
Use new TMC library

thinkyhead commented:
Looks great so far. I like the axis label being made a part of the stepper object.

teemuatlut committed:
Update config files

ejtagle commented:
The only comment i have here is that on AVR, virtual functions consume A LOT of SRAM to store virtual function tables. 
Yes, virtual function tables should be stored in FLASH, but there is a bug in the AVR compiler that stores them in SRAM, and there is no workaround for it.
I´d suggest to use a "handcrafted" virtual mechanism instead, using FLASH function pointers instead
The other choice would be to use templates as a way to implement derivation

teemuatlut commented:
I don't know much about the subject, is there somehow I can measure the impact the virtual functions have on memory consumption?

ejtagle commented:
@teemuatlut : Just compile with and without TMC support and see the amount of free SRAM. You will be suprised! .. (In fact, i was surprised the first time i tested this.. :( ) ... On 32bit processors, there is no difference at all, as virtual function tables are stored on FLASH ... Just count the amount of storage bytes your class requires. You will notice that adding a class with virtual methods consumes much more space than the one you are using. Even defining a class that inherits and overrides just one virtual function, but adds no data members, will consume a lot of SRAM (just try it! ;) )


AVR does not store virtual function tables on FLASH: Each derived class requires a virtual function table, and all of them are stored in SRAM in AVR.

That is the reason there are no virtual functions on the Marlin codebase!

AVR SRAM is very little, so every possible effort to reduce its consumption should be done 👍 
You can get mostly the same effect as virtual functions with templates...

ejtagle commented:
Please look at
https://stackoverflow.com/questions/16988450/c-what-is-the-curiously-recurring-template-pattern-and-can-curiously-recurri

You will get even better performance without the SRAM penalty of virtual functions!

teemuatlut commented:
I'll definitely go through your link. Thanks!

I remember I did originally try some other approaches than inheritance and virtual functions but ran into some issues or just didn't know how to write something.

Perhaps it's not best to compare with and without TMC enabled, as of course with a new feature enabled, and a somewhat substantial as this, the memory consumption will increase.
So what I did was compare commits from current bugfix-2.0.x branch and after the commits introduced in this PR. The current TMC2130Stepper library makes no use of virtual functions.
Here's what I got with up-to-date PIO:

```
Typical 4x TMC2130 setup on AVR

Old lib

DATA:    [====      ]  37.6% (used 3080 bytes from 8192 bytes)
PROGRAM: [==        ]  22.4% (used 56894 bytes from 253952 bytes)

New lib

DATA:    [====      ]  36.9% (used 3026 bytes from 8192 bytes)
PROGRAM: [==        ]  22.8% (used 57944 bytes from 253952 bytes)

Vanilla Marlin

DATA:    [===       ]  31.0% (used 2540 bytes from 8192 bytes)
PROGRAM: [==        ]  20.9% (used 53018 bytes from 253952 bytes)
```

To me this seems like a small reduction in SRAM use. Maybe we can conclude that the old library is even worse then :D

Also comparing to vanilla Marlin with no configuration changes, the penalty for enabling the TMC2130 drivers is +6% in SRAM. I think I'm okey with this.

I'll try testing a 32bit platform once I get back home.

ejtagle commented:
@teemuatlut : If the previous library also had virtual functions or was using or inheriting from classes with virtual functions, that could explain a bit.
I could be wrong, please correct me if i am, but, what are the requirements of static storage of the TMC library ? (i am referring to globals, or per instance data, not local variables ? - I think they probably are non existant or quite small. 
Adding the TMC library is using (from vanilla Marlin) 3026 - 2540 = 486 bytes of globals? ... I think it is pretty high ... ! - That would essentially allow to store 120 bytes per axis! - I even doubt the TMC chips have so many internal registers...
I´d suggest inspecting the .MAP file created by the linker...
The problem with SRAM is the Planner. The more available memory we have, the more we can plan ahead movements (and that helps with stutter ... longer planned movement queue means less chances of it becoming empty, specially important with small segments or UBL...

I am a big fan of templates and templated functions: They have no runtime penalties, they compile to optimal code, If you look at the STL sources, you will notice there are no virtual methods or classes...
In AVR, the indirect call to virtual functions not only costs SRAM, but also costs execution time, as it is not straightforward to perform an indirect call
(On 32 bits, that cost decreases to nearly nothing... AVR is the problem .. ;)

teemuatlut commented:
> If the previous library also had virtual functions or was using or inheriting from classes with virtual functions, that could explain a bit.

([TMC2130Stepper](https://github.com/teemuatlut/TMC2130Stepper/blob/master/src/TMC2130Stepper.h) 'old library')
No inheritance, no virtual functions. Only one you could consider is an SW SPI object created within the class.

There are 26 shadow register per driver/stepper/axis. Each is 32bits.
The old library also stores EN, DIR, STEP and CS pins as 16bit variables.
Additionally there's a 16bit value for storing the user configured RMS input and a boolean flag for using SW SPI.
Then there's a stored sense resistor value, a started flag, a tracked otpw flag and a variable for storing a status response byte.
Currently there are also some variables for storing values needed by the LCD implementation but those will be moved to within Marlin.

The SW SPI stores some precalculated fast io registers so they don't have to be looked up on every write. These are 3x16b + 3x8b + 3x8b.

There are no static variables used.

All this is per one instance of the class. By my calculations this is 1136b = 142B. Multiply this by the number of axis and you'll get 568 bytes.

Looking at it like this, there definitely is room for optimization but I don't want to start changing a working architecture to chase after 10 or 20 bytes.

Concerning templates, I think if you look around in Marlin you'll see that I'm pretty much the only one who has made use of them =)

ejtagle commented:
Yes, @teemuatlut ... templates could be helping quite a bit on Marlin ... Perhaps instead of the millions of macro lines ;) ...
Yes, i do agree that SRAM consumption matches. 
My experience with virtual functions on AVR was always terrible. Specifically the Stream class from Arduino has a huge virtual function table, and everything seems to inherit from it...

SJ-Innovation commented:
Following this: started my own rewrite but realised how much of a job it was going to be. Really like how the labels are held to each stepper instance. Thats very VERY useful.

teemuatlut commented:
@thinkyhead I've merged in your PR and squashed it in with the original config file update commit.
I also took the opportunity to disable sensorless Z by default.

thinkyhead commented:
Probably virtual methods can be eliminated from the new TMC library. _"If you always know what the real type of an object is going to be at runtime, then you don't need virtual functions."_

thinkyhead commented:
Probably virtual methods can be eliminated from the new TMC library. _"If you always know what the real type of an object is going to be at runtime, then you don't need virtual functions."_

I had a look at the `TMCStepper` library and I think you can basically just drop all the pure virtual methods completely, and you can get rid of the virtual `rms_current` methods too. As long as the types of the `TMCStepper` subclassed objects are known at compile time, and as long as they are stored in a way that preserves their class, no virtual methods should be needed.

I don't see anyplace where `TMCStepper` subclassed objects are stored as `TMCStepper*` so I think everything should continue to work as-is with all virtual methods stripped out.

thinkyhead commented:
So, I did a first attempt to throw away the virtuals, and of course it's not that simple.

The `TMCStepper` class is calling the `virtual` methods from within itself, which of course requires the virtual routing to get to the correct subclass method.

So now I'm wondering if there's another technique to get to the right subclass method. For example, by using a helper class with methods that take the different object pointer types as the first argument. I'll do some experimentation soon and see what I can come up with.

teemuatlut commented:
Virtual functions were my solution when a parent class needs to call a method from a child class.
I believe this is what virtual functions were made for. What would my alternative be?

`Read` and `write` were made virtual because the implementation differs from driver type to another, and the parent class `TMCStepper` needs to call the right one.
The same pretty much applies to the rest of the virtual methods. They are all called by a parent class method but the definitions themselves are part of a child class.

So for example, the `TMCStepper::blank_time()` method needs to call either `TMC2130Stepper::tbl()` or `TMC2208Stepper::tbl()`, depending on which child class created the parent.
When I search for "How to call child method from parent" virtual functions seem to be the appropriate answer. Of course the discussion doesn't normally include mentions of AVR.

teemuatlut commented:
> of course it's not that simple

😄 Quoted for emphasis.

I actually spent quite a few weeks trying to come up with the proper architecture and this is what worked for me.
And I'd still like to point out that even if we were to get it working and in a sensible manner, it still needs have reasonably significant memory savings to be worth the trouble.

thinkyhead commented:
I've been thinking about possible solutions. The first idea that comes to mind is to use a custom router in the root class, replacing all the methods that are currently pure virtual…
```cpp
void TMCStepper::write(const uint8_t addr, const uint32_t config) {
  switch (this->classID) {
    case TMC_2130_CLASS:
      ((TMC2130Stepper*)this)->write(addr, config);
      break;
    case TMC_2208_CLASS:
      ((TMC2208Stepper*)this)->write(addr, config);
      break;
    case TMC_2660_CLASS:
      ((TMC2660Stepper*)this)->write(addr, config);
      break;
    case TMC_5130_CLASS:
      ((TMC5130Stepper)this)->write(addr, config);
      break;
    case TMC_5160_CLASS:
      ((TMC5160Stepper*)this)->write(addr, config);
      break;
  }
}
```
While this does cost some PROGMEM, it costs zero SRAM.

thinkyhead commented:
I've been thinking about possible solutions. The first idea that comes to mind is to use a custom router in the root class, replacing all the methods that are currently pure virtual…

<details><summary>Code Samples…</summary>

```cpp
void TMCStepper::write(const uint8_t addr, const uint32_t config) {
  switch (this->classID) { // Populated by constructors (see below)
    case TMC2130_CLASS: static_cast<TMC2130Stepper*>(this)->write(addr, config); break;
    case TMC2208_CLASS: static_cast<TMC2208Stepper*>(this)->write(addr, config); break;
    case TMC2660_CLASS: static_cast<TMC2660Stepper*>(this)->write(addr, config); break;
    case TMC5130_CLASS: static_cast<TMC5130Stepper*>(this)->write(addr, config); break;
    case TMC5160_CLASS: static_cast<TMC5160Stepper*>(this)->write(addr, config); break;
  }
}
```
(…which can be shortened to…)
```cpp
#define ROUTE(T) break; case TMC##T##_CLASS: static_cast<TMC##T##Stepper*>(this)

void TMCStepper::write(const uint8_t addr, const uint32_t config) {
  switch (this->classID) { default:
    ROUTE(2130)->write(addr, config);
    ROUTE(2208)->write(addr, config);
    ROUTE(2660)->write(addr, config);
    ROUTE(5130)->write(addr, config);
    ROUTE(5160)->write(addr, config);
  }
}
```

While this does cost some PROGMEM, it costs only 1 extra byte of SRAM per object, for the `classID`.

```cpp
enum TMCClassID : unsigned char {
  TMC2130_CLASS,
  TMC2208_CLASS,
  TMC2660_CLASS,
  TMC5130_CLASS,
  TMC5160_CLASS
};

class TMCStepper {
  TMCClassID classID;
  . . .
  TMCStepper(TMCClassID cid, float RS) : classID(cid), Rsense(RS) {};
}
```

```cpp
TMC2130Stepper::TMC2130Stepper(uint16_t pinCS, float RS) :
  TMCStepper(TMC2130_CLASS, RS),
  _pinCS(pinCS)
  {}
```
</details>

ejtagle commented:
Templates! ... You are not thinking in templates.. Let me show a way (it is NOT the only way, just a way:

```cpp
class Base {
public:
  void performTask() {
    method1();
    method2();
  }

  virtual void method1() {}
  virtual void method2() {}
};

class Derived : public Base {
public:
  virtual void method1() { printf("Derived method1"); }
  virtual void method2() { printf("Derived method2"); }
};
```
Replacing virtual by templated base class:

```cpp
template <class DerivedClass>
class Base {
  void performTask() {
    self().method1();
    self().method2();
  }
  
protected:
  DerivedClass& self() { return *static_cast<DerivedClass*>(this); }
  const DerivedClass& self() const { return *static_cast<const DerivedClass*>(this); }
};

class Derived : public Base<Derived> {
public:
  void method1() { printf("Derived method1"); }
  void method2() { printf("Derived method2"); }
}
```

The ONLY reason for using virtual classes is if you want to have an array with pointers to the Base class! And, on AVR, i´d even
avoid it... How ? -- Implementing "virtual function table by hand"

```cpp
typedef void (*pfnMethod1)(void);
typedef void (*pfnMethod2)(void);

struct VirtualTable {
  pfnMethod1 method1;
  pfnMethod2 method2;
};

static void myMethod1() {
}

static void myMethod2() {
}

VirtualTable vtbl PROGMEM = {
  myMethod1,
  myMethod2
};
```

thinkyhead commented:
Here's a little C++ program that demonstrates the routing technique:

[router.cpp.txt](https://github.com/MarlinFirmware/Marlin/files/2373421/router.cpp.txt)

thinkyhead commented:
> The ONLY reason for using virtual classes is if you want to have an array with pointers to the Base class!

Or, if you want to do a call of a method from the root class, but that method is only promised in the root class (pure virtual) and only implemented in the derived classes.

```cpp
TMCStepper::do_something(int x, int y) {
  my_method(123, y, 456); // this method is implemented differently by each derived class
}
```

thinkyhead commented:
Ah, I see that your `self()` method in the templated `Base<>` class accomplished the same thing, but neater.

thinkyhead commented:
Ah, I see that your `self()` methods in the templated `Base<>` class accomplish the same thing, but in a much neater way. So, following that technique we get this much simpler looking thing!

[templated.cpp.txt](https://github.com/MarlinFirmware/Marlin/files/2373444/templated.cpp.txt)

I'm now a much bigger fan of `static_cast`.

ejtagle commented:
The sweetness of templates is that all the calls are resolved at compile time.. No wasted cycles, no wasted SRAM or resources :)

thinkyhead commented:
Funny, because I've employed this technique before… A looong time ago. It's pretty much a necessity in many cases to have a `self()` to get to the method you want. I've had to do this in C# and JavaScript also.

thinkyhead commented:
Are templated sub-classes also possible without extending this? For example, `TMC5130Stepper` is a subclass of `TMC2130Stepper`. So we need to be able to do this…
```cpp
class TMC5130Stepper : public TMC2130Stepper<TMC5130Stepper> {
  . . .
};
```

ejtagle commented:
Should be possible.
static_cast will always work for single inheritance, just because C++ just adds the derived class members in the same order as the inheritance (base class type ... outmost derived class type) - So the address of the base class is the same as the address of the derived classes

thinkyhead commented:
It's tricky. This of course doesn't work at all for obvious reasons…
```cpp
template <class T>
class TMC2130Stepper : public TMCStepper<T> { . . . }

template <class T>
class TMC5130Stepper : public TMCStepper<T> { . . . }

class TMC5160Stepper : public TMC5130Stepper<TMC5160Stepper> { . . . }
```

ejtagle commented:
(and yes, you can achieve the results you want by making TMC2130Stepper a template itself ... There are easy ways to make it behave

```cpp
template <class T>
class TMC2130Stepper  { };
```

A `typedef` can "remove" the parameters and give an even cleaner usage pattern if you want that.

ejtagle commented:
```cpp
template <class T>
class TMC2130Stepper : public TMCStepper<T> { . . . };

template <class T>
class TMC5130Stepper : public TMCStepper<T> { . . . };
```

That should work ... Why not ? ...

thinkyhead commented:
Yes, you're correct. Since `TMC5130Stepper` is a subclass of `TMC2130Stepper`, as long as `self()` casts to the end of the chain (`TMC5130Stepper`) the calls to `self().do_something()` will get to the right (non-virtual) method, whether it's implemented in one or both sub-classes.

thinkyhead commented:
But here's a problem… What if we want to create an instance of `TMC2130Stepper`?
How does `TMCStepper` get a `T` that equals `TMC2130Stepper`?

`TMC2130Stepper<TMC2130Stepper> stepperX;` ??

ejtagle commented:
Yes, that is a "syntax" problem, not a big problem. You must always pass to the base class template parameter the derived class type
So, the answer is .. ;)
```
template <class T>
class TMCStepper {
}

template<class T>
class TMC2130Stepper: public TMCStepper { ... }

TMC2130Stepper<TMC2130Stepper> stepper; 
```

Remember, we are passing the type, not an instance of the object ;)

ejtagle commented:
Ok, now i get it... you need to specify the type of the template

thinkyhead commented:
So to correct your example:
```cpp
template <class T>
class TMCStepper { . . . };

template<class T>
class TMC2130Stepper: public TMCStepper<T> { . . . };

TMC2130Stepper<TMC2130Stepper> stepper; 
```

Syntax ugliness indeed. So is it best if they are ALL template classes? Otherwise we would have `TMC5160Stepper` and `TMC2224Stepper` as exceptions.
```cpp
  TMC2130Stepper<TMC2130Stepper> stepperX;
  TMC5160Stepper stepperY;
  TMC2224Stepper stepperZ;
  TMC5130Stepper<TMC5130Stepper> stepperE0;
```
 (`TMC2660Stepper` is not a sub-class of `TMCStepper` but a root class of its own.)

ejtagle commented:
You are right @thinkyhead ... My example will not compile. The cause is that you cannot use as a type a template without specifying the type of its parameters

```
TMC2130Stepper<TMC2130Stepper> stepper;  
```

But you can do:

```
class TMC2130StepperFinal : public TMC2130Stepper<TMC2130StepperFinal> {};
TMC2130StepperFinal stepperX;
```

thinkyhead commented:
Yeah, I was thinking that some intermediate classes might be needed to allow simpler instantiations.

ejtagle commented:
Maybe renaming the classes will look nicer:
```
template <class T>
class TMCStepper { };

template <class T>
class TMC2130StepperBase : public TMCStepper<T> { };

class TMC2130Stepper : public TMC2130StepperBase<TMC2130Stepper> {};

template <class T>
class TMC5130StepperBase : public TMCStepper<T> { };

class TMC5130Stepper : public TMC5130StepperBase<TMC5130Stepper> {};

```

thinkyhead commented:
Yeah, I was just doing exactly that. But guess why it doesn't work.

ejtagle commented:
Should work, let me try it ;)

ejtagle commented:
Works for me :) 
```

#include <iostream>
using namespace std;

template <class T>
class TMCStepper { 
public:
	TMCStepper() {}
	~TMCStepper() {}
	inline void calls_routed_method() { self().routed_method(123); }

protected:
	T& self() { return *static_cast<T*>(this); }
	const T& self() const { return *static_cast<const T*>(this); }
};

template <class T>
class TMC2130StepperBase : public TMCStepper<T> {
public:
	TMC2130StepperBase() {}
	~TMC2130StepperBase() {}
	void routed_method(int x);
};

class TMC2130Stepper : public TMC2130StepperBase<TMC2130Stepper> {};

template <class T>
class TMC5130StepperBase : public TMCStepper<T> { 
public:
	TMC5130StepperBase() {}
	~TMC5130StepperBase() {}
	void routed_method(int x);
};

class TMC5130Stepper : public TMC5130StepperBase<TMC5130Stepper> {};

template<class T>
void TMC2130StepperBase<T>::routed_method(int x) {
	cout << "TMC2130 routed_method " << x << endl;
}

template<class T>
void TMC5130StepperBase<T>::routed_method(int x) {
	cout << "TMC5130 routed_method " << x << endl;
}


int main()
{
	TMC2130Stepper stepperX;
	TMC5130Stepper stepperY;
	stepperX.routed_method(234);
	stepperX.calls_routed_method();
	stepperY.routed_method(234);
	stepperY.calls_routed_method();
	return 0;
}

```

thinkyhead commented:
I'm stuck at this:
```
In file included from ../Projects/Maker/Firmware/TMCStepper/examples/Calibrate_spreadCycle/Calibrate_spreadCycle.ino:5:0:
../Documents/Arduino/libraries/TMCStepper/src/TMCStepper.h: In member function 'void TMC5130StepperBase<T>::diag1_poscomp_pushpull(bool)':
../Documents/Arduino/libraries/TMCStepper/src/TMCStepper.h:349:46: warning: there are no arguments to 'self' that depend on a template parameter, so a declaration of 'self' must be available [-fpermissive]
   void diag1_poscomp_pushpull(bool B) { self().diag1_pushpull(B); }
                                              ^
```

ejtagle commented:
Well...
C++ idiosincracies ...

try ```this->self()``` instead of ```self()``` ...

thinkyhead commented:
That works. Next issue! The `using myclass::mymethod` statement requires that `myclass` be a parent class of the class where `using` is used. So if two siblings are inheriting from a common parent class they can't do `using` with each other.
```
Documents/Arduino/libraries/TMCStepper/src/TMCStepper.h:611:25:
error: type 'TMC5130Stepper' is not a base type for type 'TMC5160Stepper'
   using TMC5130Stepper::I_scale_analog;
```

ejtagle commented:
Yes, but if a class requires to use the method of a sibling, then something is wrong ... Maybe move the method to a common ancestor, maybe duplicate it, maybe... There are several alternatives here... (but still this is something very strange...)

ejtagle commented:
and the issue itself of using a sibling will not work unless it is an static method. This will not even work if using virtual functions...

ejtagle commented:
In the TMCStepper library, you should just keep the class hierachy as it is, and it will work. 
Assuming you define 
```
template<class T>
class TMC5130StepperBase {} 
```
then 
```
template<class T>
TMC5160StepperBase : public TMC5130StepperBase<T> { ... }
```

And replace
```
using TMC5130Stepper::I_scale_analog;
```

by
```
using TMC5130StepperBase<T>::I_scale_analog;
```

and it should work 👍

thinkyhead commented:
### Old inheritance chain:

- `TMCStepper`
  - `TMC2130Stepper`
    - `TMC5130Stepper`
      - `TMC5160Stepper`
  - `TMC2208Stepper`
    - `TMC2224Stepper`
- `TMC2660Stepper`

### New Inheritance Chain
- `TMCStepper<T>`
    - `TMC2130StepperBase<T>`
      - `TMC2130Stepper`
      - `TMC5130StepperBase<T>`
        - `TMC5130Stepper`
        - `TMC5160Stepper`
    - `TMC2208StepperBase<T>`
      - `TMC2208Stepper`
      - `TMC2224Stepper`
- `TMC2660Stepper`

thinkyhead commented:
Here's the work so far, as a PR against the library: https://github.com/teemuatlut/TMCStepper/pull/6

And here's the branch if you want to check it out and try some things: https://github.com/thinkyhead/TMCStepper/tree/templatize_classes

At this point, the `using` statements are blocking compilation when building the `Calibrate_spreadCycle.ino` example sketch. I haven't gotten to the point where it will start to complain about method calls that need `this->self().` in front of them. That will be the next piece.

ejtagle commented:
There was nothing preventing you to keep 
```
TMC2130StepperBase<T>
	TMC5130StepperBase<T>
```
👍 

All the rest are mostly the same.

thinkyhead commented:
Just corrected that.

ejtagle commented:
BTW, great work! ... On AVR my experience is always the same - virtual functions are a show stopper, and must be avoided if possible. On ARM, i have no problems with them...

thinkyhead commented:
So now… How to get `TMC5160Stepper` to be a child of `TMC5130Stepper`? I see a possible infinite regress issue looming.

ejtagle commented:
Why not? ...
Example...
```
template <class T>
class TMC5160StepperBase : public TMC5130StepperBase<T> {}

class TMC5160Stepper : public TMC5160StepperBase <TMC5160Stepper> {};
```

Probably you are already seeing the pattern. Always use the -Base, and create a "Usable" or "Final" class for the one the user will use

BTW, this will also fix the 'using' problems ;)

thinkyhead commented:
OK, solved that. New problem!

~~Template method definitions / implementations must be in headers.~~

…n/m just need `template<typename T>` prefixes on all method definitions.

ejtagle commented:
yes, but not necesarily.. You can leave template method definitions in cpp files if you explicitly instantiate the templates (https://stackoverflow.com/questions/115703/storing-c-template-function-definitions-in-a-cpp-file)

In your .h file...
```
template<typename T>
class foo
{
public:
    void bar(const T &t);
};
```
And in your .cpp file
```
template <class T>
void foo<T>::bar(const T &t)
{ }

// Explicit template instantiation
template class foo<int>;
```

teemuatlut commented:
@ejtagle How could I determine the size of vtable? Preferably without rewriting the library to be able to compile without virtual methods.

ejtagle commented:
@teemuatlut : theoretically, for each virtual function (compute virtual functions as all different (named) functions that have the virtual attribute in all the class hierachy... ) . 2 bytes per each function. On 32bit targets, it should be 4 bytes...
That is all. The other option is to look at the MAP file and search for 'vtable for'  and you will find the count of bytes of a vtable for the mentioned class

teemuatlut commented:
I think I got the file but it doesn't mention "vtable" at all.
[output.map.txt](https://github.com/MarlinFirmware/Marlin/files/2373928/output.map.txt)

Added compiler flags were: `-Xlinker  -Wl,-Map,output.map`

ejtagle commented:
Yes, that map file is not verbose... let me check ...

ejtagle commented:
Humm... on AVR seems due to global optimizations, the vtable symbol is not displayed on the MAP file... :(

ejtagle commented:
(https://gcc.gnu.org/bugzilla/show_bug.cgi?id=43745) ... this is the bug on avr-gcc related to vtables...

thinkyhead commented:
Ok, so I converted everything into template-y goodness and pushed it all up to https://github.com/teemuatlut/TMCStepper/pull/6. At this point the build of `Calibrate_spreadCycle.ino` is almost completing, but I'm getting link errors… This is quite a puzzler.

![image](https://user-images.githubusercontent.com/698003/45410821-50833500-b638-11e8-9402-3e6621bd7832.png)

ejtagle commented:
That is a link error: Probably there is a missing explicit template instantiation.
When the linker complains a given templated function is not found, that is because it was not instantiated by the compiler, or maybe the definitions of those functions were not visible to the compiler when the template was explicitly instantiated  :)

teemuatlut commented:
I've made some memory optimizations with the TMCStepper library.
I removed some _unnecessary_ virtual functions, moved register addresses to progmem and modified some of the read only registers to not have a shadow register as that is not actually needed.
For a full set of 4x TMC2130 I'm currently looking at SRAM consumption of 334 bytes.
This is down 206 bytes compared to current upstream Marlin.

oldmcg commented:
Is the goal to offer MIXED stepper driver support in single build/config for 8-bit controller?

Or should 8-bit target be limited to single model of stepper driver, meaning no virtual calls necessary (just plumb with appropriate #defines)?

Just food for thought.

ejtagle commented:
@teemuatlut : That is great! ... I did finish the templatized version (not tested!) based on the work already done y @thinkyhead ... Maybe you can grab the best of both worlds... ;) 
https://github.com/ejtagle/TMCStepper , on the templatize_classes branch.

334 bytes is not bad at all 👍

teemuatlut commented:
> Is the goal to offer MIXED stepper driver support in single build/config for 8-bit controller?

Yes. This is already possible and I definitely do not want to take it away. I couldn't be printing myself if I did that!

> Maybe you can grab the best of both worlds...

Ultimately for me the question is if I want to move to a (much) more complicated system to save (13 x 16b) 26 bytes.
I will have to test what the final difference is between the two.

ejtagle commented:
@teemuatlut : Templated library is not "more complex"... it is just different... Virtual functions are not "simpler" ... For example, with virtual functions you can end calling a pure virtual function if you are not careful enough... 
It took me a little bit of time to get used to templates, but once i did, i´d never go back... The optimizations that are achievable with templates simply can´t be achieved with virtual functions (compiler is unable to inline virtual functions, unless it can probe at compile time the specific virtual function to call)
Yes, in this library there are no "speed constrains", so it is not critical at all, i agree

thinkyhead commented:
> I will have to test what the final difference is between the two.

I'm very curious to see the difference in code size and SRAM usage. Although there is some added complexity, I think it's the right approach. Old-school "embedded C++" doesn't even permit `virtual`.

One thing we may also be able to use without overhead is STL. I've seen some videos describing how to use it effectively, and it can produce some surprisingly small assembly code.

THIS 👉 https://www.youtube.com/watch?v=zBkNBP00wJE 👈

ejtagle commented:
STL, if properly used (even if i don´t like its resulting syntax!) can cost nearly nothing. It depends on how you use it

thinkyhead commented:
Everyone who programs C++ for embedded systems _**MUST watch the linked video**_. It's mind-blowing. Also, the online tools used in the video for previewing the (x86) assembler output are invaluable. We should be using that site, or something like it.

p3p commented:
Jason Turners videos in general are must watches unless you have the time to actually parse the new standards completely, there are numerous features of c++14 - 17 that make embedded programming clean (death to the preprocessor ;) ) and efficient.

ejtagle commented:
In the case of this library, i´d expect increased flash usage, and decreased SRAM usage... Maybe the linker will be smart enough and collapse all similar functions in just one... i don´t know.
But, the power of c++ is that if you give enough information to the compiler via types, constexprs and templates, it can perform extremely smart optimizations at compile time, No other language does that kind of optimization.
It is said that C++ compiles all the compilers of the other languages, with reason ;)

teemuatlut commented:
Question: how do you make an abstract class without a virtual function?

thinkyhead commented:
@teemuatlut — The same way as you make cake without flour. 🤣 Which is to say, an "abstract class" is defined by having only pure virtual functions.

thinkyhead commented:
Here are the results of a build of `Calibrate_spreadCycle.ino`:

Build size before: 11252 / 402
Build size after: 9000 / 362

Code is 2252 bytes smaller. Data is 40 bytes smaller.

thinkyhead commented:
@teemuatlut — The same way as you make cake without flour. 🤣 Which is to say, an "abstract class" is defined by having only pure virtual functions.

I think an abstract class can be emulated using templates, basically by following the same kind of design pattern we're following with this templatized conversion. The main difference is that the C++ language won't force derived classes to implement the pure virtuals.

---
Here are the results of a build of `Calibrate_spreadCycle.ino`:

Build size before: 11252 / 402
Build size after: 9000 / 362

Code is 2252 bytes smaller. Data is 40 bytes smaller.

teemuatlut commented:
Have you pushed the latest commits online yet?

thinkyhead commented:
Yep! All the latest stuff is pushed and visible at https://github.com/teemuatlut/TMCStepper/pull/6

p3p commented:
> But, the power of c++ is that if you give enough information to the compiler via types, constexprs and templates, it can perform extremely smart optimizations at compile time

I especially like when previewing assembly built from complex STL structures and templates, arrays etc using that tool and it gets optimised to a constant at runtime, I always assume constexpr now when I write functions, until I'm forced not to.

teemuatlut commented:
Getting this when building Marlin:
```
Linking .pioenvs\megaatmega2560\firmware.elf
c:/users/teemu/.platformio/packages/toolchain-atmelavr/bin/../lib/gcc/avr/4.9.2/../../../../avr/bin/ld.exe: COOLCONF.cpp.o: plugin needed to handle lto object
c:/users/teemu/.platformio/packages/toolchain-atmelavr/bin/../lib/gcc/avr/4.9.2/../../../../avr/bin/ld.exe: COOLCONF.cpp.o: plugin needed to handle lto object
C:\Users\teemu\AppData\Local\Temp\cc1t9MNb.ltrans7.ltrans.o: In function `tmc_init(TMC2130Stepper&, unsigned int, unsigned int, unsigned long, float) [clone .constprop.21]':
cc1t9MNb.ltrans7.o:(.text+0x380): undefined reference to `TMC2130StepperBase<TMC2130Stepper>::COOLCONF(unsigned long)'
collect2.exe: error: ld returned 1 exit status
*** [.pioenvs\megaatmega2560\firmware.elf] Error 1
```

>  it gets optimised to a constant at runtime

At compile time maybe? I hope? 😃 

> Templated library is not "more complex"

Templates themselves may not be complicated, but I'm more referring to code like this:
```cpp
T& self() { return *static_cast<T*>(this); }
const T& self() const { return *static_cast<const T*>(this); }
```
This *most definitely* is more complex than what was before =)

ejtagle commented:
@teemuatlut : That is a avr-gcc bug 'plugin needed to handle lto object' 
(https://forum.arduino.cc/index.php?topic=415801.0) -- But it doesn´t happen to me...

ejtagle commented:
Yes, the ```self()``` function is the "compile time" "virtual function" implementation ;)

thinkyhead commented:
Apparently Arduino IDE 1.8.7 is out with some compiler fixes.

marcio-ao commented:
This thread is interesting. I ran into the same issue with virtual functions when developing the graphical UI for our printers. I ended up building my own virtual function table using pointers stored in PROGMEM. If you want to see my technique, [look here](https://github.com/marcio-ao/Marlin/blob/bugfix-2.0.x-for-lulzbot-archim2-with-touchscreen-and-usb/Marlin/src/lcd/extensible_ui/lib/ui_framework.h).

Depeding on what virtual functions are needed for, another option is the [curiously recurring template pattern](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern). I have little experience with that technique, but it seems clever.

teemuatlut commented:
Status update on my end.
I am now down to ~~75~~ 71 bytes of SRAM used when compiling the `Simple` example. 12 bytes out of that is used by the SPI library. By my calculations, the member variables take 59 bytes and cannot be avoided.
I left virtual functions for `read` and `write`. They also serve to abstract the `TMCStepper` class.
Only `TMCStepper` is a class template. Other functions were done with your static_cast monster.

*The virtual functions take a total of 4 bytes.*

https://github.com/teemuatlut/TMCStepper/tree/dev

ejtagle commented:
(@marcio-ao : I did exactly the same for a Graphical UI i was creating for (Marlin?) sometime ago. It is not complete yet, as i am working in the antialiased font rendering (the custom compressor/decompressor i wrote for those fonts reduce them to manageable sizes: a full proportional 16x16 antialiased font takes 3kbytes) ... 
The color bitmap (image) compressor/decompressor is done (compression ratio similar to PNG, but just 300 bytes of SRAM, and decompression speed is extremely fast)
Graphics programming always takes time, specially when you strive for high quality graphics and there are not enough resources ...)

marcio-ao commented:
@ejtagle: Fortunately the graphics chip we are using (the FTDI FT810) has fonts and most widgets built in. Very little use of resources to get a fully functional GUI.

ejtagle commented:
@teemuatlut : If you don't need to store pointers to the base class TMCStepper from where all the other classes inherit, then you -actually- don´t need virtual functions at all ,,,
The "cost" of those virtual functions you are using is:
2 bytes per function * number of classes that inherit from TMCStepper: 

2 * 2 * 6 = 24 bytes of SRAM, and 24 bytes of FLASH (this is worst case, assuming someone instantiates all the classes that inherit from TMCStepper)

Trust me, unless you need to collect pointers to the TMCStepper base class, you don´t need virtual functions at all, and it is a very interesting and useful talent to learn how to avoid them if unneeded 
 - STL is a great example of programming with no virtual functions at all ... ;)

ejtagle commented:
(@marcio-ao : My target was/is a dumb controller based on the ILI93xx chipset. LCDs using that chip are extremely cheap ... You would be surprised of the results achievable with it. (BTW, it is just a hobby project of mine... i do not like the idea of having 2 controllers, one for graphics, one for control, because it adds complexity to the setup... without significant gains ... To use a graphics coprocessor is not justifiable unless you are doing 3d renderings of the model that you are printing, or something like that ...;) )

marcio-ao commented:
@ejtagle: I know this is getting off topic, but if you are developing a GUI, I would invite you to take a look at the extensible_ui interface I proposed in PR#11193. The idea is to have a well-defined API so that the user can drop in different types of graphical UIs. If you have any feedback on that, it would be appreciated.

> To use a graphics coprocessor is not justifiable unless you are doing 3d renderings of the model that you are printing, or something like that ...

Would an 8-bit micro be able to refresh an 800x480, 24-bit color display? I sort of doubt it. The FTDI chip helps by doing the heavy-lifting for that.

ejtagle commented:
@marcio-ao : An 8bit micro can at most perform 1 refresh per second on such displays, but most UIs do not require such whole display refresh speeds and/or do not require whole screen refresh at all (just small parts of it). The main problem, in my personal opinion, with u8glib is that it insists on performing whole screen refreshes, even if you just change a small part of the screen.
If i was on a budget, what i´d do is to throw away the 8bit AVR controller, use a 32bit controller instead (i am using a Due, but several others are even faster!) and use a cheaper LCD color display.
A 32bit controller is more than capable of a decent UI running concurrently with Marlin, would give improved control of the printer and using "spare cycles" can keep the display updated.
Do you want even more for the same price ? ... Use an ESP32 module as 3d printer controller, driving a color LCD though SPI. You get an incredibly fast system for nearly nothing... !

marcio-ao commented:
> most UIs do not require such whole display refresh speeds and/or do not require whole screen refresh at all

@ejtagle: It depends on what quality you are looking for. I've always intensely disliked the DuePanel, so if that's an example what can be done with a 32-bit board, I am not impressed at all. All the ones I have seen redraw from left to right at a fairly sluggish pace and the boxy widgets look like a the original X Windows widgets from the 80s. The UI [we are working on for our printers](https://github.com/MarlinFirmware/Marlin/issues/10008) is a huge step up, but someday I hope there is an open-source UI that looks as gorgeous as what the [Raise3D printers](https://robosavvy.com/store/raise-3d-n2-plus-dual.html) are using.

ejtagle commented:
@marcio-ao : No, i am working with a parallel version of the ILI93xx . Update speed is pretty good. Right now i am aiming at an ui similar to Raise3D... Not impossible at all. 
Once i get this whole thing debugged and working, i´ll post some pictures...

marcio-ao commented:
> Once i get this whole thing debugged and working, i´ll post some pictures...

@ejtagle: Sounds good! Looking forward to seeing it :)

thinkyhead commented:
Rebased this concept onto the updated `TMCStepper` library and got the following results:

### `Simple.ino` (TMC2130 Software)
- Before: 5550 / 107
- After: 4652 / 75 (-898, -32)

### `Software_SPI.ino`
- Before: 6800 / 331
- After: 6290 / 299 (-610 / -32)

### `Calibrate_spreadCycle.ino`
- Before: 9440 / 346
- After: 8932 / 314 (-508 / -32)

### `Live_tune.ino`
- Before: 14954 / 1095
- After: 14372 / 1063 (-582, -32)

teemuatlut commented:
Current dev branch:

|                  | SRAM | PROGMEM
|------------|--------|-------------
| Simple       |  71     | 4510
| SW_SPI      | 305    | 6302
| Calibrate   | 320    | 8760
| Live tune   | 1062  | 14372

thinkyhead commented:
@teemuatlut — I've rebased this and fixed it up for the Z3 axis and E5 stepper. But I'm unable to push it to your branch. ~~If you could do a rebase of your branch onto the latest `bugfix-2.0.x` and push the rebase up to your fork, then I will be able to submit the differences to your fork as a PR.~~

thinkyhead commented:
Better yet, you can just grab the rebased code here: https://github.com/thinkyhead/Marlin/tree/bf2_new_library_rebase

teemuatlut commented:
Rebased to upstream.

brandstaetter committed:
Support for manually deployed fixed probes (#11899)

thinkyhead commented:
Rebased, and had to move to a new PR… #11919

teemuatlut commented:
I disabled edits so I could review any additional changes needed.

thinkyhead commented:
The rebase was complicated by some dead commits. It needed a refresh. You can pull the HEAD of the new PR’s branch in as the new HEAD of your branch to rectify it, using a `git reset`.

teemuatlut commented:
I hard reseted to your branch and rebased it to upstream.
I then added some new commits. Can you review the last 5 commits and I can then sort & rebase them into something sensible. I thought it would be easier for you to go through them before I combined the new things into old commits.
You may need to re-open this PR, or [look at the branch directly](https://github.com/teemuatlut/Marlin/commits/bf2_new_library). [Commit diff](https://github.com/teemuatlut/Marlin/compare/cac669b5e63d8eb1176898c40490b25d23d9dd73...a17c0448b347879c1b04bff4f18313c095c9bf74).

Here are the current standings in terms of memory consumption:

| Config (4x TMC2130)                        | PROGMEM       | SRAM |
|:----------------------------------------|-----------------|--------|
| Upstream Marlin with TMC disabled | 52192              | 2499  |
| TMC enabled                                     | 56002              | 2999  |
| New library                                        | 55594              | 2752 (-247B) |

So with this PR, basic TMC2130 support should consume only 253 bytes of SRAM (down from 500B).

ejtagle commented:
@teemuatlut  : Make TMCStorage a template also, and instead of passing a pointer to the label, you can pass the label as a template parameter.
Very easy, and can save 4 extra bytes per axis 👍 

(using SRAM to store non changing pointers or values is a waste of resources ;) )

BTW: This TMCLibrary is in great shape... I see no other optimizations possible right now. Great work!!

teemuatlut commented:
I'll take a look.
The storing of the label was supposed to be a somewhat direct port from before but I guess at some point it did turn into a run time variable.

teemuatlut commented:
Feel free to pitch in. Getting too late for me.
This is what I'm trying to test with:
```cpp
static const char LABEL[] PROGMEM = "MSG_A";

template<char L>
class myClass {
	public:
	void foo() { char* c = L; }
};

void setup() {
  myClass<LABEL> c1;
  c1.foo();
}
```
```cpp
CPP_template_label:13:11: error: the value of 'LABEL' is not usable in a constant expression
CPP_template_label.ino:4:19: note: 'LABEL' was not declared 'constexpr'
CPP_template_label:13:16: error: conversion from 'const char*' to 'char' not considered for non-type template argument
```

ejtagle commented:
@teemuatlut : This works for me... the only requirement is to make the string constexpr

```
static constexpr char demo [] PROGMEM = "DEMO";

template <const char* p> class myClass {
public:
	static const char* text() { return p; }
};

int main()
{
	myClass<demo> test;
	printf("%s", test.text());
}
```

This works for me on both AVR and DUE

teemuatlut commented:
Okey, the next step that I need is passing the object as a argument to a function. I don't know why it's complaining as it's exactly the same as I had in Marlin.
The compiler message isn't helping much either 😄 
`expected a constant of type 'const char*', got 'const char*'`

```cpp
static constexpr char LABEL[] PROGMEM = "MSG_A";

template<const char* L>
class myClass {
	public:
	void foo() { const char* c = L; (void)c; }
};

void fun(myClass<const char*> &st) { (void)st; }

void setup() {
  myClass<LABEL> c1;
  c1.foo();
  fun(c1);
}
```

ejtagle commented:
That one is "easy"...

The template you are using requires not a type, but a value to be able instantiate it...

This will work, but probably is not what you need:

```cpp
static constexpr char LABEL[] PROGMEM = "MSG_A";

template<const char* L>
class myClass {
  public:
  void foo() { const char* c = L; (void)c; }
};

void fun(myClass<LABEL> &st) { (void)st; }

void setup() {
  myClass<LABEL> c1;
  c1.foo();
  fun(c1);
}

```

ejtagle commented:
One possible solution to this is to use

```cpp
template <typename T> void fun(T &st) { (void)st; }
```

ejtagle commented:
Another possible solution is 
```cpp
template <const char* T> void fun(myClass<T> &st) { (void)st; }
```

This latter one is less generic, but i think it is the one you need @teemuatlut

teemuatlut commented:
Well, it compiles but doesn't link. For some reason Marlin lost the definition for stepperX.

https://github.com/teemuatlut/Marlin/commit/0f0da30874dcb3620c382a33eed287b22cbbbbfa

```cpp
extern TMCMarlin<TMC2130Stepper, TMC_X_LABEL> stepperX;

TMCMarlin<TMC2130Stepper, TMC_X_LABEL> stepperX(X_CS_PIN, R_SENSE);
```

ejtagle commented:
Wow! ... Let me see ... ;) ...

ejtagle commented:
@teemuatlut : This seems to be a bug of GCC 4.8.x / 4.9.x series when using LTO (linker time optimizations) that sometimes removes global symbols from modules when it shouldn't (https://answers.launchpad.net/gcc-arm-embedded/+question/261316) , (https://trac.mplayerhq.hu/ticket/1695).
I am trying to find a workaround for it, but until now, none worked...

thinkyhead commented:
> I hard reseted to your branch and rebased it to upstream.

I don't see that this PR was rebased, or new commits added. And now it cannot be re-opened, so you may want to just open a new PR based on your current work branch.

teemuatlut commented:
I don't think github keeps tracking PRs that were closed. If you follow the links to the branch commit history, you'd see plenty of new commits.

thinkyhead commented:
Probably. If I had re-opened it sooner then it would have them. Anyway, go ahead and start a new PR so you can have control over its content.
